### PR TITLE
Issue-35: Death handling and combat end

### PR DESCRIPTION
**Pull request by Tomzkk**
# Description

Handle enemy deaths and remove them from combat, as well as detect the end of combat and emit a signal
Currently also reloads the TestingScene on end of battle

## Related issue(s)

https://github.com/Saplings-Projects/1M_sub/issues/35

## List of changes

A more detailed list of the changes.
- Update Battler.gd to detect and handle enemy deaths as well as detect and emit signal on end of combat
- Refactored enemy attack loop in Battler.gd to two separate loops
    - First generate enemy actions into a separate list, then execute them in order
    - Created class EnemyAction which holds information on who is trying to do what action on what target
- Added a scene controller which can reload a scene - used to restart combat when signal is detected
- Created a signal in PhaseManager which is used to indicate end of battle (emitted from Battler and connected to SceneController)
- Changed setter in PartyComponent to set the array rather then extend it

## Tests

Currently none

## Additional notes

- I made SceneController just so that we can see combat actually end and restart and play with it. I imagine actual management of scenes etc will probably be working differently. Currently scene is completely reloaded whether player wins or loses for simplicity

### CODE REVIEW

File: Core/Battler.gd
Comment (Turtyo) [inline Core/Battler.gd:75]: i think `_handle_deaths() should be called after each enemy's turn start status has been applied.
Imagine the following situation:
Two enemies have 1 HP, each has a dot of poison. One enemy heals other enemies when it dies, by 3 HP, let's say enemy A
- If enemy A dies then enemy B, enemy A heals B then B gets his poison dot applies and survives
- If enemy B dies then enemy A, both die
- In the current situation, since we add all the enemies to the list of enemies to remove, both enemies die in each case because they are removed before actions take effect

I don't know if this is preferable, maybe we want to keep it as is so the order of the enemies doesn't influence situations

File: Core/Battler.gd
Comment (Turtyo) [inline Core/Battler.gd:157]: this TODO should be done also

File: Core/SceneController.gd
Comment (Turtyo) [inline Core/SceneController.gd:30]: name is not explicit, missing typing

File: Core/SceneController.gd
Comment (Turtyo) [inline Core/SceneController.gd:7]: missing typing

File: Core/Battler.gd
Comment (Turtyo) [inline Core/Battler.gd:133]: missing typing for enemy (same for all `for` loops)

File: Core/SceneController.gd
Comment (Turtyo) [inline Core/SceneController.gd:12]: missing `void` return type

File: Core/SceneController.gd
Comment (Turtyo) [inline Core/SceneController.gd:25]: missing `void` return type

File: Core/SceneController.gd
Comment (Turtyo) [inline Core/SceneController.gd:42]: missing `void` return type

File: Core/SceneController.gd
Comment (Turtyo) [inline Core/SceneController.gd:46]: instead of putting the name directly here, I think using a dictionary with `{DEFEAT_SCENE: defeat_scene_path, WIN_SCENE: win_scene_path` will allow us to change the dictionary in the future without having to change this function or other functions that do the same

File: Global/EnemyAction.gd
Comment (Turtyo) [inline Global/EnemyAction.gd:9]: might be better to use different names from the class names to not be confused

File: Global/EnemyAction.gd
Comment (Turtyo) [inline Global/EnemyAction.gd:7]: this should not be named target (singular) when it's an array potentially containing multiple targets
`list_targets` might be better

File: Core/SceneController.gd
Comment (Turtyo) [inline Core/SceneController.gd:4]: missing typing

**Review by Turtyo — CHANGES_REQUESTED** [review]
The general structure is clear (thank to the earlier explanation in the issue). There are some problems with the code which I commented. Thank you for the PR, I think it will work nicely in the end :+1: 
Tests have to be provided before this can be merged

File: Core/Battler.gd
Comment (Tomzkk) [reply Core/Battler.gd:26]: I am reusing the _"on_card_action_finished"_ signal which is defined in CardManager and used for other purposes. This also comes with an argument for a card that's being played. I just need to know that a card has been played and do not care for which card at this moment, so I opted for unbinding it rather then adding an argument to the __handle_deaths_ function that would never be used. 

I could also make my own signal without an argument and use it for this if you think it would be better to keep the signals separate for some reason. I didn't want to create a signal which would basically perfectly shadow the _on_card_action_finished_ signal we already had as I didn't see a reason to. Let me know if you can think of any concerns regarding that

Comment (Turtyo) [reply Core/Battler.gd:26]: No no it's fine as is, I just wanted to make sure to understand

File: Core/Battler.gd
Comment (Tomzkk) [reply Core/Battler.gd:75]: A fair point, although if I wanted to call it after each enemy's turn I would need to do same thing as I did for their attacks as it is within loop through _enemy_list.

Actually, in most games I know, statuses like poison etc are only resolved when it comes to specific enemies' turn, however that would require us to rework the enemy turn to have them first be hit with status effects, then attack and only then hand the turn over to the next enemy. 
Currently, statuses as poison etc are specifically resolved as "on turn start" events, which implies they should be handled right at the start of the turn start
 
If we wanted to rework it into what i mentioned above, I am not sure how much work that would require - maybe it would want to be a separate issue
If we want to keep them as is and be resolved at the start of the turn, then I'm not sure which order would be preferred 🤔

File: Core/Battler.gd
Comment (Tomzkk) [reply Core/Battler.gd:157]: Currently, I emit a signal and SceneController completely reloads the scene - this means entities get reloaded from 0 and their stats are gone. As such the todo doesn't really apply as the entity which holds the stats gets recreated per my understanding - which is why I left the todo in place so that it gets implemented once proper handling of combat progression is implemented

File: Core/SceneController.gd
Comment (cheesefrycook) [reply Core/SceneController.gd:46]: I agree - in the future we could have other scenes in the dictionary too like the Shop and Rest Area

File: Global/EnemyAction.gd
Comment (cheesefrycook) [inline Global/EnemyAction.gd:1]: How does RefCounted differ from Resource?

File: Core/SceneController.gd
Comment (cheesefrycook) [inline Core/SceneController.gd:43]: So I'm guessing PhaseManager's `_ready` doesn't get called whenever we reload the scene which is why we need to call it manually here. Would it be problematic to move the logic from the `_ready` of PhaseManager into a new function and then call that function manually after the scene is loaded?

File: Core/Battler.gd
Comment (cheesefrycook) [reply Core/Battler.gd:75]: Maybe it helps to separate the status events into two separate cases?
`on_turn_start` = when the `ENEMY_ATTACKING` phase starts
`on_enemy_before_action` = before the specific enemy's action. Only applied to the enemy that is casting its action.

Also, #38 will make each enemy attack one after another. So Enemy A attacks, plays an animation. After some delay, Enemy B attacks and plays its animation.

File: Global/EnemyAction.gd
Comment (Tomzkk) [reply Global/EnemyAction.gd:1]: https://docs.godotengine.org/en/stable/tutorials/best_practices/node_alternatives.html

Per my understanding the main difference is Resource adds the ability to be serialized/deserialized which allows you to export properties into/from the inspector. If I'm not misunderstanding or missing something then I do not really think we need anything from Resource specifically ?

File: Core/SceneController.gd
Comment (Tomzkk) [reply Core/SceneController.gd:48]: I don't really see a problem with it in PhaseManager. 
We might not even need to touch phases at the start of this function and switching the phases at the end of it (by calling the function like current PhaseManager's ready) might be enough ?
 
I think the reason I did it like this was I wanted to switch to a Phase that doesn't have any events happening inside of it to prevent things happening while the scene is being cleaned and reloaded - although I'm not sure if that is actually necessary or even fulfills what I intended 🤔

File: Global/EnemyAction.gd
Comment (cheesefrycook) [reply Global/EnemyAction.gd:1]: Yeah, I guess if we don't need any export properties then this is fine

File: Core/SceneController.gd
Comment (cheesefrycook) [reply Core/SceneController.gd:48]: I'm thinking it may be sufficient to switch phases after the load is complete. However if we run into problems and want to prevent stuff from happening during the load process, we could potentially add a new phase like `GAME_END` or something

File: Core/Battler.gd
Comment (Turtyo) [reply Core/Battler.gd:157]: The reset stat will reset the temp stats, which need to be reset anyway
Doing this would allow for proper testing of stat reset at combat end (though I already tested them in unit tests)

It also makes sense to leave it as is for when we have multiple combats I guess

File: Core/SceneController.gd
Comment (Turtyo) [reply Core/SceneController.gd]: exactly as Cheesy said

File: Core/SceneController.gd
Comment (Turtyo) [reply Core/SceneController.gd:48]: `GAME_END` would be confusing, `SCENE_END` probably better
but anyway, when you load a new scene it will call the tree that will await for ready signals from it's children nodes, so I think it should all setup properly when reloading a new scene :thinking:

File: Tests/test_death_handling.gd
Comment (Turtyo) [inline Tests/test_death_handling.gd:63]: should also have a function that does when player is killed by an enemy attack

File: Tests/test_death_handling.gd
Comment (Turtyo) [inline Tests/test_death_handling.gd:104]: can you make a test that uses damage all enemies to test that we can kill both enemies through a single attack ?

File: Tests/test_death_handling.gd
Comment (Turtyo) [inline Tests/test_death_handling.gd:112]: should also make a function where the duration of the poison is set to 1 so poison expires when enemy dies
this would allow to check everything works fine on the order in which things are done

File: Core/Battler.gd
Comment (Tomzkk) [reply Core/Battler.gd:75]: I see, I misremembered that STS does it like that then. 
The effects that would be affected by playing them in that order are mainly damage debuffs - so that if an enemy has a damage over time on it that would do enough damage to kill it, other enemies still have a window to do something about it
Though ultimately I think it's okay either way, as long as it's consistent

The only ways I see it doable right now are basically redoing what I did with enemyAction (albeit simpler) and splitting the debuff loop - so I would have to duplicate enemy list and apply statuses while going through it rather then through the original - or by making debuff handling part of EnemyAction's execution (which would however also change when debuffs are handled)

File: Core/Battler.gd
Comment (Tomzkk) [reply Core/Battler.gd:157]: Per my understanding temp stats are still bound to the entity (so the player in our case I believe), which gets deleted and recreated since I reload the scene, or am I wrong there ? If that is the case, why is there a need to reset temp stats ?

File: Global/EnemyAction.gd
Comment (Tomzkk) [reply Global/EnemyAction.gd:9]: I believe underscored variables are in most cases used for instance variables, so it might be a bit confusing - we could also rename the function ones to be slightly different but I didn't have an idea at the time so I opted for adding underscores to the class variables

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:63]: I believe I actually had that but deleted it when I started having issues with scene reloads. Can look into adding it back

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:104]: Same as previous comment

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:130]: Added test for this one as well as for the previous two comments

File: Core/Battler.gd
Comment (Turtyo) [reply Core/Battler.gd:75]: other enemies don't "still have a window to do something about it" because turn start applications are automatic things, enemies can't do something when the turn already started to mitigate those
i think we should keep it as is, because it will be consistent no matter the order of the enemies on screen and in the list

File: Core/Battler.gd
Comment (Turtyo) [reply Core/Battler.gd:157]: Yes it gets deleted for now, but later we will copy the stats, so we will need the reset stat function to be connected to the end of combat signal anyway

As I said, it's fine to leave it as is for now, since it is indeed related to stats carrying over, so not yet implemented

File: Global/EnemyAction.gd
Comment (Turtyo) [reply Global/EnemyAction.gd:9]: In GDScript convention, variables using an underscore are "private" and are not meant to be accessed from outside the scope of the current script, just called by this script and children of this script
Thus it would be more logical to use `¢aster`, `ŧarget` and `target_list` for the class variables and `_caster`,`_target` and `_target_list` for the `_init` variable
As you can see, `_ready` functions and `_init` functions are a good example of that. They use an underscore because you don't call them directly (`_init` gets called by new() and `_ready` gets called when the tree is building the nodes)

File: Tests/test_death_handling.gd
Comment (Turtyo) [inline Tests/test_death_handling.gd:148]: does that even work ? isn't the enemy dead thus the resource is freed and you can't check his status component ?

File: Core/SceneController.gd
Comment (Tomzkk) [reply Core/SceneController.gd:48]: What do we think the action here should be then ? 🤔 

Should I make the new "COMBAT_END / SCENE END / ..." phase, switch to it at the beginning of this and then call the equivalent of _ready function of phaseManager after the switch like @cheesefrycook suggested ? Or ?

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:148]: freeing happens deferred via queue_free so it doesn't actually get freed until the next frame - so at the end of the test the enemy is still a valid node and we can check it - at least that's my understanding and the test has been consistently passing but let me know if you think I'm wrong in something

Actually I don't even see a way to check if enemy has actually really been freed in a test (which I wanted to assert for as well) because I would have to await for the next frame in the test - which I tried to do but it then broke things as it seemed to me that the "_before_each_" was executed when the next test was already in progress and it freed the node midrun

File: Global/EnemyAction.gd
Comment (Tomzkk) [reply Global/EnemyAction.gd]: Yes, I'm saying the same thing as you are - it's just that the underscore in name is overall usually reserved for instance variables or methods that are private rather then for function arguments. Even the examples you mentioned are class functions (_init, _ready). Although I'm not sure how closely godot follows it, but from what I've seen it does seem to be the case

Overall it's just a convention though and not a big problem as the init should be pretty obvious anyway, so I swapped them around 👍

File: Core/SceneController.gd
Comment (Turtyo) [reply Core/SceneController.gd:48]: Well imo I think it should work without that
Meaning you shouldn't have to set the `GAME_STARTING` for the game to be in this specific phase, as I explained it should naturally be set like that when the scene starts

File: Global/EnemyAction.gd
Comment (Turtyo) [reply Global/EnemyAction.gd]: well for now you have class variables with an underscore, (meaning private), and function parameters without an underscore (meaning public)
I would expect it to be other way around

File: Tests/test_death_handling.gd
Comment (Turtyo) [reply Tests/test_death_handling.gd:148]: `before_each` is executed at the start of each test, so if you await in a specific test, it shouldn't just jump to the next test and launch the `before_each`

`_on_enemy_start_turn` calls `_handle_deaths` which calls `_handle_enemy_deaths`, and in there you add everything to a list to remove, erase and queue free. `queue_free` actually frees at the end of the current frame. If that's the only thing you need to wait you can either:
- `await get_tree().process_frame` to wait for the deferred calls
- `is_queued_for_deletion()` doesn't wait for deferred calls but you can check the enemy is indeed waiting to disappear
Option one is probably better as it makes sure everything works instead of assuming the `queue_free` works directly

File: Core/SceneController.gd
Comment (Tomzkk) [reply Core/SceneController.gd:48]: PhaseManager is autoloaded, so just reloading the scene doesn't do anything with phases. 
If I remove the phase settings I am doing here, what happens is enemy turn finishes running (as scene reload is deferred) which switches phase to PLAYER_ATTACKING. After the reload, the phase stays on PLAYER_ATTACKING. The issue with this is that the cards are not served to the player (as cards are served when it detects change from some phase into PLAYER_ATTACKING) - game gets into a state where it is player's turn but he has no cards and can only end turn

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:148]: When I try to do it via option one, which is essentially the same as I did before which I described above, it results in tests throwing error "Nonexistent function in base 'Nil'" at some point during test's execution - which the only way I can see happening is if the player was freed in the middle of the test run and the only way we are freeing it is at the end of the previous test 🤔

Option 2 works fine

File: Core/Battler.gd
Comment (Tomzkk) [reply Core/Battler.gd:75]: What I meant was if we resolved those things at the start of enemy action, there could be enemy that would have it's turn before affected enemy which could cleanse it / heal it / whatever.

File: Core/SceneController.gd
Comment (Turtyo) [reply Core/SceneController.gd:48]: Aaaaah i forgot that PhaseManager is autoloaded, you are right ! So you indeed need to call the `_ready`
I think doing a `SCENE_END` phase is probably the best here then

File: Tests/test_death_handling.gd
Comment (Turtyo) [reply Tests/test_death_handling.gd:148]: Did you try to put some breakpoints in godot to check when things are being freed ? That might allow you to see why option one doesn't work

File: Core/Battler.gd
Comment (Turtyo) [reply Core/Battler.gd:75]: ah yes I see
we will keep it as is anyway I think, because it's more consistent that way

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:148]: It seems the issue is PhaseManager is being recreated at a wrong time somehow and switches phase to Player turn which tries to apply player statuses at a time when player doesn't exist
image


Also it looks like the issue is happening if I use both await inside of tests AND queue_free in _after_each_. I tried switching the freeing after tests to _free_ and it seems like it's passing without issues

Comment (Turtyo) [reply Tests/test_death_handling.gd:148]: well as long as it's passing consistently it's fine

**Review by Turtyo — APPROVED** [review]

**Comment by Turtyo** [comment]
Well I think there is only the comment for the `SCENE_END` phase that needs to be done and it should all be good

**Review by cheesefrycook — APPROVED** [review]

File: Core/SceneController.gd
Comment (Tomzkk) [reply Core/SceneController.gd]: I wasn't sure if it would be better to do SCENE_END or call the _ready at the end (rather then just changing phase to player's turn) - in the end I did both. This way the SCENE_END is currently redundant I believe, but makes it more clear what's happening and might have use in the future ? [See here](https://github.com/Saplings-Projects/1M_sub/pull/60/files#diff-c6159dde6e380cb517c924628c3146eff48ca4e4b4d2cdc89af1c5adf9290290R47-R55)

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:168]: Added asserts into tests - I kept both queued_for_deletion before await and checking if node was freed after

**Review by cheesefrycook — APPROVED** [review]

File: Tests/test_death_handling.gd
Comment (Turtyo) [inline Tests/test_death_handling.gd:53]: missing typing, you removed it when it was already there; any reason ?

**Review by Turtyo — CHANGES_REQUESTED** [review]

File: Core/SceneController.gd
Comment (Turtyo) [reply Core/SceneController.gd]: yes i read it, it's clear what happens; a good idea I think :+1:

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:53]: If the node is invalid (as it was freed when the enemy died), it is no longer a node type. 

I don't think godot has option for multi-typing and even if it did, I'm not sure what type would even cover invalid node if any 🤔

Comment (Turtyo) [reply Tests/test_death_handling.gd:53]: mmh I see the problem indeed;
could you put a comment so we don't forget about this exception ?

File: Tests/test_death_handling.gd
Comment (Turtyo) [inline Tests/test_death_handling.gd:1]: after pulling the branch, I realized that this test doesn't use the structure that was made in #63 
Normal, since this PR was started first, but could you update this to extend from `TestBase` ? You will be able to remove all variable declarations, most of the `before_each` (except after the `watch_signals`, but you can call `super` for everything before)
You will need to leave the the `after_each`, and could you also add an `assert_no_new_orphans` as in done in other tests ? This is to check for runtime errors mainly (cf #51 and #62 )

**Review by Turtyo — CHANGES_REQUESTED** [review]
Something to change in test that I noticed by pulling the branch. I think we'll be good to merge after this is done :+1:

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd:16]: Added

File: Tests/test_death_handling.gd
Comment (Tomzkk) [reply Tests/test_death_handling.gd]: Done

**Review by Turtyo — APPROVED** [review]

**Review by cheesefrycook — APPROVED** [review]

**Pull request merged**
